src/Compilers/CSharp/Test/Semantic/Semantics/CollectionLiteralTests.cs

```diff
@@ -523,14 +523,50 @@ static void Main()
                 """;
             var comp = CreateCompilation(source);
             comp.VerifyEmitDiagnostics(
-                // (5,13): error CS0815: Cannot assign collection literals to an implicitly-typed variable
+                // (5,17): error CS9503: There is no target type for the collection literal.
                 //         var x = [null, 1];
-                Diagnostic(ErrorCode.ERR_ImplicitlyTypedVariableAssignedBadValue, "x = [null, 1]").WithArguments("collection literals").WithLocation(5, 13),
+                Diagnostic(ErrorCode.ERR_CollectionLiteralNoTargetType, "[null, 1]").WithLocation(5, 17),
                 // (6,20): error CS9500: Cannot initialize type 'object' with a collection literal because the type is not constructible.
                 //         object y = [null, 2];
                 Diagnostic(ErrorCode.ERR_CollectionLiteralTargetTypeNotConstructible, "[null, 2]").WithArguments("object").WithLocation(6, 20));
         }
 
+        [Fact]
+        public void NaturalType_24()
+        {
+            string source = """
+                class Program
+                {
+                    static void Main(string[] args)
+                    {
+                        (string, int)[] x1 = [(null, default)];
+                        string[] y1 = [args.Length switch { 0 => null, _ => default }];
+                        string[] z1 = [args.Length == 0 ? null : default];
+                        var x2 = [(null, default)];
+                        var y2 = [args.Length switch { 0 => null, _ => default }];
+                        var z2 = [args.Length == 0 ? null : default];
+                    }
+                }
+                """;
+            var comp = CreateCompilation(source);
+            comp.VerifyEmitDiagnostics(
```

Comment:
Please pull on the semantic info in a way that will cause a crash if it survives initial binding unconverted, such as IOperation.